### PR TITLE
libs: Sort Genesis on file SaveAs

### PR DESCRIPTION
**Pull request by jackzampolin**
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by jackzampolin** [comment]
Related to https://github.com/cosmos/cosmos-sdk/pull/4294

File: types/utils.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+// MustSortJSON is like SortJSON but panic if an error occurs, e.g., if
+// the passed JSON isn't valid.
+func MustSortJSON(toSortJSON []byte) []byte {
+	js, err := SortJSON(toSortJSON)
+	if err != nil {
+		panic(err)
+	}
+	return js
+}
```
Comment (Liamsi) [inline types/utils.go]: This is basically copy & pasted from the SDK (added here: https://github.com/cosmos/cosmos-sdk/commit/b7f902fba5fcb367d82a2dd9e59600e91d2dfc72#diff-da7a1d1eef56dc7672476c992ce1e56c) 

I think we should also copy over the tests then. Probably these methods (and tests) do not need to live in both repositories (but that is up to the SDK team).

Comment (melekes) [reply types/utils.go]: Should we move them to `libs/common` package?

Comment (Liamsi) [reply types/utils.go]: Yes, that is probably a good place for this to live in.

Comment (jackzampolin) [reply types/utils.go]: I'm super OK with not duplicating code. I didn't want to import the SDK into tendermint tho... How should I move forward here?

Comment (Liamsi) [reply types/utils.go]: I think we should move those methods into libs/common together with their tests. Whenever the sdk updates to a Tendermint Version including this PR, they could delete the code from the SDK.

Comment (Liamsi) [reply types/utils.go]: https://github.com/tendermint/tendermint/pull/3650

**Comment by melekes** [comment]
What's the rationale behind this change?

**Comment by jackzampolin** [comment]
@melekes During the last network hardfork we ran into an issue where when exporting, and then generating the genesis file we were getting different hashes. This would sort the JSON keys by default and should help avoid a similar issue in the future.

File: libs/common/json_sort.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package common
```
Comment (melekes) [inline libs/common/json_sort.go]: I'd just call the file `json.go`. If you look at other files, you'll see they are named similarly - `int.go`, ...

File: types/genesis.go [hunk 3]
```diff
@@ -47,7 +47,7 @@ func (genDoc *GenesisDoc) SaveAs(file string) error {
 	if err != nil {
 		return err
 	}
-	return cmn.WriteFile(file, genDocBytes, 0644)
+	return cmn.WriteFile(file, cmn.MustSortJSON(genDocBytes), 0644)
```
Comment (melekes) [inline types/genesis.go]: I think we should make this explicit by adding `sort bool` parameter.

Comment (melekes) [reply types/genesis.go]: Also, "SaveAs is a utility method for saving GenensisDoc as a JSON file." comment needs to be updated

**Comment by melekes** [comment]
> During the last network hardfork we ran into an issue where when exporting, and then generating the genesis file we were getting different hashes.

Because you did not sort `app_state`? I don't understand. The order should always be the same for `GenesisDoc` struct keys when encoded as json:

```
genesis_time
chain_id
consensus_params
validators
app_hash
app_state
```

**Pull request closed without merging**